migrations/04_singleton_proxy.js

```diff
@@ -0,0 +1,104 @@
+
+const SingletonFactory = artifacts.require("SingletonFactory")
+const WitnetRequestBoard = artifacts.require("WitnetRequestBoard")
+
+module.exports = async function (deployer, network, accounts) {
+
+  let addresses = require('./addresses.json')
+
+  const factory = await SingletonFactory.deployed()
+  const wrb = await WitnetRequestBoard.deployed()
+
+  const singletons = require('./singletons.json') 
+  const artifact = artifacts.require("WitnetRequestBoardProxy")
+  const contract = artifact.contractName
+
+  if (singletons.contracts[contract]) {
+
+    let addresses = require('./addresses.json')
+
+    const from = singletons.from
+        ? singletons.from
+        : deployer.networks[network].from
+          ? deployer.networks[network].from
+          : accounts[0]
+      ;
+
+    const salt = singletons.contracts[contract].salt
+        ? '0x' + ethUtils.setLengthLeft(ethUtils.toBuffer(singletons.contracts[contract].salt), 32).toString('hex')
+        : '0x0'
+      ; 
+
+    let bytecode = artifact.toJSON().bytecode
+    if (singletons.contracts[contract].links) singletons.contracts[contract].links.forEach(
+      // Join every dependent library address into the contract bytecode to be deployed:
+      lib => {
+        const lib_artifact = artifacts.require(lib)
+        const lib_addr = lib_artifact.address.slice(2).toLowerCase()
+        const lib_mark = `__${lib_artifact.contractName}${'_'.repeat(38 - lib_artifact.contractName.length)}`
+        bytecode = bytecode.split(lib_mark).join(lib_addr)
+      }
+    )
+    artifact.bytecode = bytecode
+
+    var contract_addr = await factory.determineAddr.call(bytecode, salt)
+
+    if ((await web3.eth.getCode(contract_addr)).length <= 3) {
+      // Deploy contract instance, if not yet deployed on this `network`:
+      traceHeader(`Singleton inception of contract '${contract}':`)
+
+      const balance = await web3.eth.getBalance(from)
+      const gas = singletons.contracts[contract].gas
+          ? singletons.contracts[contract].gas
+          : 10 ** 6
+        ;
+
+      // Compose initialization call: 'upgradeWitnetRequestBoard(wrb.address)'
+      let initCall = '0x47b1e79b000000000000000000000000' + wrb.address.slice(2)
+
+      const tx = await factory.deployAndInit(bytecode, initCall, salt, {from: from, gas: gas})
+      utils.traceEvents(tx.logs)
+      traceDeploymentTx(tx.receipt, web3.utils.fromWei((balance - await web3.eth.getBalance(from)).toString()))
+
+    } else {
+      traceHeader(`Singleton contract: '${contract}'`)
+    }
+
+    artifact.address = contract_addr
+    const contract_codehash = web3.utils.soliditySha3(await web3.eth.getCode(artifact.address))
+    if (!contract_codehash) {
+      throw new Error(`Fatal: unable to deploy contract '${contract}' as singleton. Try providing more gas.`)
+    }
+
+    utils.logs.trace("  ", "> contract codehash:  ", web3.utils.soliditySha3(await web3.eth.getCode(artifact.address)))
```

Comment:
Why not simply use `console.log`?

Reply:
`Utils` also provides a way to print debug traces (`utils.logs.debug`) that
may get silenced, or reactivated, by modifying utils.js::_debug. Besides,
utils is meant to centralize other helper functions that may become handy
for either tracing or performing migrations scripts (i.e. traceEvents,
shortenAddr, generateDeployTx, rawTransactions).

Reply:
Just asking because I saw that you could silence `debug` log messages but not `trace` (which are lower). Therefore,my first thought was...maybe just use `console.log`. :smile:

Reply:
Another posibility is to use a logger library such as `js-logger`

Reply:
It also seems strange to me to implement our own logger whose main advantage over plain `console.log` is silencing one log level instead of allowing to chose the minimum level to be shown. I'd rather go with any off-the-shelf logging solution.

Reply:
Okay, it was just a simple and fast implementation for just-one intended purpose. If you agree, I could just substitute all `utils.logs.trace` for `console.log`, by now. If later on logs categorization and/or filtering were needed, we could then switch to `js-logger`.

Reply:
That's a great take for the time being. Thanks!